### PR TITLE
[#722] More clearly explain the async nature of aliases

### PR DESCRIPTION
**Pull request by richdouglasevans**
Addresses #722 

screen shot 2018-08-24 at 08 20 03

### CODE REVIEW

File: source/api/commands/as.md [hunk 1]
```diff
@@ -1,9 +1,8 @@
 ---
 title: as
-
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: I noticed that this extra newline was being removed in recent PRs so I duly removed it here too for consistency.

File: source/api/commands/as.md [hunk 2]
```diff
@@ -1,9 +1,8 @@
...
 ---
 
-Assign an alias for later use. Reference the alias later within a {% url `cy.get()` get %} or {% url `cy.wait()` wait %} command with a `@` prefix.
+Assign an alias for later use. Reference the alias later within a {% url `cy.get()` get %} or {% url `cy.wait()` wait %} command with an `@` prefix.
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: `@` is pronounced _"at"_ so the "a" becomes an "an."

> Such alliteration, much wow 😊

File: source/api/commands/as.md [hunk 3]
```diff
@@ -21,9 +20,9 @@ Assign an alias for later use. Reference the alias later within a {% url `cy.get
 
 ```javascript
 cy.get('.main-nav').find('li').first().as('firstNav') // Alias first 'li' as @firstNav
-cy.route('PUT', 'users', 'fx:user').as('putUser')     // Alias 'route' as @putUser   
-cy.stub(api, 'onUnauth').as('unauth')                 // Alias 'stub' as @unauth   
-cy.spy(win, 'fetch').as('winFetch')                   // Alias 'spy' as @winFetch  
+cy.route('PUT', 'users', 'fx:user').as('putUser')     // Alias that route as @putUser   
+cy.stub(api, 'onUnauth').as('unauth')                 // Alias that stub as @unauth   
+cy.spy(win, 'fetch').as('winFetch')                   // Alias that spy as @winFetch
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: This just read a wee bit oddly to my eyes so I changed it... we're not aliasing each function itself, we're aliasing what is yielded by each functions. That's why I added the _"that"_ qualifier.

File: source/api/commands/as.md [hunk 4]
```diff
@@ -36,7 +35,7 @@ cy.as('foo')   // Errors, cannot be chained off 'cy'
 
 **{% fa fa-angle-right %} aliasName** ***(String)***
 
-The name of the alias to be referenced later within a {% url `cy.get()` get %} or {% url `cy.wait()` wait %} command using a `@` prefix.
+The name of the alias to be referenced later within a {% url `cy.get()` get %} or {% url `cy.wait()` wait %} command using an `@` prefix.
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: Such alliteration, much wow... Part II 😊

File: source/api/commands/as.md [hunk 5]
```diff
@@ -46,22 +45,44 @@ The name of the alias to be referenced later within a {% url `cy.get()` get %} o
...
+
 ```javascript
-cy.route('PUT', /^\/users\/\d+/, 'fixture:user').as('userPut')
-cy.get('form').submit()
-cy.wait('@userPut')
-  .its('url').should('contain', 'users')
+it('disables on click', function () {
+  cy.get('button[type=submit]').as('submitBtn')
+  cy.get('@submitBtn').click().should('be.disabled')
+})
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: The original example wasn't actually aliasing a DOM element 😕so I changed it to one that does actually alias a DOM element.

File: source/api/commands/as.md [hunk 6]
```diff
@@ -46,22 +45,44 @@ The name of the alias to be referenced later within a {% url `cy.get()` get %} o
 
 ## DOM Element
 
+Aliasing a DOM element and then {% url getting get %} that aliased element.
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: Style guide steer Jennifer: is linking to commands using phrases or plain words such as _"getting"_ for `cy.get()` cool? The sentence reads smoother in this style but I can't recall seeing an example of this style elsewhere in the docs.

File: source/api/commands/as.md [hunk 7]
```diff
@@ -46,22 +45,44 @@ The name of the alias to be referenced later within a {% url `cy.get()` get %} o
...
+
+```javascript
+beforeEach(function () {
+  cy.fixture('users-admins.json').as('admins');
+});
+
+it('the users fixture is bound to this.admins', function () {
+  cy.log(`There are ${this.admins.length} administrators.`);
+});
+```
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: There wasn't an example of aliasing a fixture so I added one.

File: source/api/commands/as.md [hunk 8]
```diff
@@ -46,22 +45,44 @@ The name of the alias to be referenced later within a {% url `cy.get()` get %} o
...
+
+{% note warning %}
+Note the use of the standard function syntax. Using {% url 'arrow functions' https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions %} to access aliases via `this` won't work because of {% url 'the lexical binding' https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#No_separate_this %} of `this`.
+{% endnote %}
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: Here and elsewhere in this PR I've started hyperlinking more frequently. I noticed that you -- Jennifer -- have started doing this in recent PRs so I'm following your lead here. I dunno if hyperlinking to the MDN docs as an authoritative source is cool: steer me if not. I find the MDN docs to be good reference material.

File: source/api/commands/as.md [hunk 9]
```diff
@@ -70,6 +91,47 @@ cy.wait('@userPut')
 
 Some strings are not allowed as alias names since they are reserved words in Cypress. These words include: `test`, `runnable`, `timeout`, `slow`, `skip`, and `inspect`.
 
+## `as` is asynchronous
```
Comment (richdouglasevans) [inline source/api/commands/as.md]: A wee section that gives more explicit examples to address #722.

It's only targeting `fixture()` aliases but I thought that adding a ton of examples for aliasing `route()` and DOM elements might be overload.

File: source/guides/core-concepts/variables-and-aliases.md [hunk 10]
```diff
@@ -58,7 +58,7 @@ cy.get('button').then(($btn) => {
 })
 ```
 
-If you're familiar with native Promises the Cypress `.then()` works the same way. You can continue to nest more Cypress commands inside of the `.then()`.
+If you're familiar with {% url 'native Promises' https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises %} the Cypress `.then()` works the same way. You can continue to nest more Cypress commands inside of the `.then()`.
```
Comment (richdouglasevans) [inline source/guides/core-concepts/variables-and-aliases.md]: I linked to the `Promise` docs on MDN in an effort to make the Cypress docs even more friendly and useful.

File: source/guides/core-concepts/variables-and-aliases.md [hunk 11]
```diff
@@ -158,7 +157,7 @@ The reason for using `const` is because the `$span` object is mutable. Whenever
 
 # Aliases
 
-Using `.then()` callback functions to access the previous command values is great - but what happens when you're running code in hooks like `before` or `beforeEach`?
+Using `.then()` callback functions to access the previous command values is great&mdash;but what happens when you're running code in hooks like `before` or `beforeEach`?
```
Comment (richdouglasevans) [inline source/guides/core-concepts/variables-and-aliases.md]: Is there a style guide for the Cypress docs somewhere Jennfier? Even an informal one? I've replaced the ` - ` with an em-dash for style... yes, no?

> I'm not that fussed... the em-dash variant looks better to my eyes.

File: source/guides/core-concepts/variables-and-aliases.md [hunk 12]
```diff
@@ -247,8 +246,8 @@ describe('parent', function () {
       })
 
       it('can access all aliases as properties', function () {
-        expect(this.a).to.eq('one') // true
-        expect(this.b).to.eq('two') // true
+        expect(this.a).to.eq('one')   // true
+        expect(this.b).to.eq('two')   // true
         expect(this.c).to.eq('three') // true
```
Comment (richdouglasevans) [inline source/guides/core-concepts/variables-and-aliases.md]: I just made 'em line up in a column... _I couldn't resist_ 😊

File: source/guides/core-concepts/variables-and-aliases.md [hunk 13]
```diff
@@ -260,7 +259,7 @@ describe('parent', function () {
 
 The most common use case for sharing context is when dealing with {% url `cy.fixture()` fixture %}.
 
-Often times you may load in a fixture in a `beforeEach` hook but want to utilize the values in your tests.
+Often times you may load a fixture in a `beforeEach` hook but want to utilize the values in your tests.
```
Comment (richdouglasevans) [inline source/guides/core-concepts/variables-and-aliases.md]: The double occurrence of the word _"in"_ tripped by tongue so I elided one of 'em: there's no loss of clarity.

File: source/guides/core-concepts/variables-and-aliases.md [hunk 14]
```diff
@@ -273,7 +272,7 @@ it('utilize users in some way', function () {
   const user = this.users[0]
 
   // make sure the header contains the first
-  // users name
+  // user's name
```
Comment (richdouglasevans) [inline source/guides/core-concepts/variables-and-aliases.md]: Adding some missing apostrophes. (Also below.)

File: source/guides/core-concepts/variables-and-aliases.md (hunk 11 above)
Comment (richdouglasevans) [inline source/guides/core-concepts/variables-and-aliases.md]: Is there a style guide for the Cypress docs somewhere Jennifer? Even an informal one? I've replaced the ` - ` with an em-dash for style... yes, no?

> I'm not that fussed... the em-dash variant looks better to my eyes but no biggie.

screen shot 2018-08-23 at 22 44 20
screen shot 2018-08-23 at 22 43 41

File: source/api/commands/as.md (hunk 6 above)
Comment (jennifer-shehane) [reply source/api/commands/as.md]: No, we have not had a use case of this before. We generally write `cy.get()` here. I would probably use `cy.get()` for more clarity - this explains the command needed without them needed to click on the actual link to see the example.

File: source/api/commands/as.md (hunk 7 above)
Comment (jennifer-shehane) [reply source/api/commands/as.md]: Interesting, yeah I alias fixtures a lot - should def have an example.

File: source/api/commands/as.md (hunk 8 above)
Comment (jennifer-shehane) [reply source/api/commands/as.md]: Yes! Awesome, we love MDN docs and are not shy about offering general good JS advice when relevant.

File: source/api/commands/as.md [hunk 15]
```diff
@@ -70,6 +91,47 @@ cy.wait('@userPut')
...
+
+Remember that **Cypress commands are async**, including `as()`.
+
+Because of this you cannot _synchronously_ access anything you have aliased. You must use other asynchronous commands such as {% url `.then()` then %} to access what you've aliased.
+
+Here are some further examples of using `as()` that illustrate the asynchronous behaviour.
```
Comment (jennifer-shehane) [inline source/api/commands/as.md]: behavior -> behavior - I know I need to write up our official style guide, but we follow US english spellings

File: source/api/commands/as.md [hunk 16]
```diff
@@ -70,6 +91,47 @@ cy.wait('@userPut')
...
+
+```javascript
+describe('A fixture', () => {
+
+  describe('alias can be accessed', () => {
+    it('via get().', () => {
+      cy.fixture('admin-users.json').as('admins');
```
Comment (jennifer-shehane) [inline source/api/commands/as.md]: We also do not use semicolons in our code examples - again, need to write up our style guide 😅

File: source/guides/core-concepts/variables-and-aliases.md [hunk 17]
```diff
@@ -86,12 +86,12 @@ cy.get(...).find(...).should(...)
...
 
 {% note info %}
-By using callback functions we've created a closure. Closures enable us to keep references around to refer to work done in previous commands.
+By using callback functions we've created a {% url closure https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures %}. Closures enable us to keep references around to refer to work done in previous commands.
 {% endnote %}
 
 ## Debugging
 
-Using `.then()` functions is an excellent opportunity to use `debugger`. This can help you understand the order in which commands are run. This also enables you to inspect the objects that Cypress yields you in each command.
+Using `.then()` functions is an excellent opportunity to use {% url `debugger` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures %}. This can help you understand the order in which commands are run. This also enables you to inspect the objects that Cypress yields you in each command.
```
Comment (jennifer-shehane) [inline source/guides/core-concepts/variables-and-aliases.md]: Was this the wrong link?

File: source/guides/core-concepts/variables-and-aliases.md (hunk 11 above)
Comment (jennifer-shehane) [reply source/guides/core-concepts/variables-and-aliases.md]: 👍

File: source/api/commands/as.md [hunk 18]
```diff
@@ -70,6 +91,47 @@ cy.wait('@userPut')
 
 Some strings are not allowed as alias names since they are reserved words in Cypress. These words include: `test`, `runnable`, `timeout`, `slow`, `skip`, and `inspect`.
 
+## `as` is asynchronous
+
+Remember that **Cypress commands are async**, including `as()`.
```
Comment (jennifer-shehane) [inline source/api/commands/as.md]: `as()` should be `.as()`

File: source/api/commands/as.md (hunk 16 above)
Comment (richdouglasevans) [reply source/api/commands/as.md]: > I need to write up our style guide.

I'm not being cheeky here... shall I write a guide and then you can ~tear it to pieces~ review it? 😄Cypress is doing great work for us and I'd like to contribute back. If me writing that would be unhelpful then no worries.

File: source/guides/core-concepts/variables-and-aliases.md (hunk 17 above)
Comment (richdouglasevans) [reply source/guides/core-concepts/variables-and-aliases.md]: Darn, I thought I fixed it when I did a self review. I will fix it in any case... huh. Ta for spotting this.

**Pull request merged**